### PR TITLE
Add FusedLayerNorm (Eigen and CUDA)

### PR DESCRIPTION
**Pull request by kaixih**
This PR adds FusedLayerNorm and FusedLayerNormGrad to the TF with Eigen and CUDA implementations for CPU and GPU respectively.

FYI: @nluehr

### CODE REVIEW

**Comment by rohan100jain** [comment]
Could you explain the motivation for a new kernel for fused layer normalization? There is already a keras layer for LayerNormalization that uses the fused batch norm for implementation https://github.com/keras-team/keras/blob/v2.7.0/keras/layers/normalization/layer_normalization.py#L29-L363

**Comment by kaixih** [comment]
Yes, actually, the next step following this PR is to plumb in this fused layer norm to the keras layer.

The current keras LayerNormalization does a "partially" fused implementation, which takes advantage of fused_batch_norm for the normalization and then a scale/offset op. The reason why we cannot wrap up all the normalization+scale/offset to fused_batch_norm is because the layer norm's var/mean has different shapes with scale/offset, whereas the fused_batch_norm requires them to be the same.

This PR implements a fully fused layer norm and our plan is to call sth like `gen_ops.fused_layer_norm` underneath the keras LayerNormalization after this gets approved.

File: tensorflow/core/kernels/fused_layer_norm_op.cc [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+                errors::InvalidArgument("input must be 2-dimensional",
+                                        x_input.shape().DebugString()));
+    OP_REQUIRES(context, scale_input.dims() == 1,
+                errors::InvalidArgument("scale must be 1-dimensional",
+                                        scale_input.shape().DebugString()));
+    OP_REQUIRES(context, offset_input.dims() == 1,
+                errors::InvalidArgument("offset must be 1-dimensional",
+                                        offset_input.shape().DebugString()));
+    const int batch_dim = 0;
+    const int feature_dim = 1;
```
Comment (mdfaijul) [inline tensorflow/core/kernels/fused_layer_norm_op.cc:276]: @kaixih Why are you restricting only 2D tensor as input? You could flatten dims [0..(n-2)] and consider as batch dimension.

Comment (kaixih) [reply tensorflow/core/kernels/fused_layer_norm_op.cc:276]: I was thinking to let the python side to reshape the tensor to 2D before calling the backend impl, since it has the `axis` value and it knows which dims should be collapsed. And I feel it might be more flexible to let the python side to handle the reshape (even transpose) for different given axis. Do you suggest we should consider the `-1` is the default feature axis and move this reshape functionality into the C++ side?

Comment (mdfaijul) [reply tensorflow/core/kernels/fused_layer_norm_op.cc:276]: @kaixih Yes, I prefer having reshape functionality into C++ side.

**Comment by mdfaijul** [comment]
@kaixih It will be great if you could add bfloat16 support for CPU. Eigen library supports several operations in bfloat16 numeric.

**Comment by kaixih** [comment]
I am wondering if we can add the bfloat16 support later as a separate PR. To support it in Eigen/CPU, it seems I need to add bfloat16 in `Attr("T: {half, float}")`. But that means the GPU also need to deal with such inputs. I am not familiar how to convert bfloat16 to float. Does it support sth like `static_cast<float>` as in the half->float?

**Comment by mdfaijul** [comment]
@kaixih Thanks for the quick reply. Yes, we can do it as separate PR.

**Comment by rohan100jain** [comment]
Adding @fchollet from Keras to comment on this as well. In general, I'm a little hesitant to add to the existing very large op set that TF already supports but if we don't have a clear alternative and this unlocks better performance, we can do this. Do you have some performance numbers or potential impact (over the existing implementation) that could justify this better?

**Comment by gbaned** [comment]
@fchollet Can you please take a look on this PR ? Thanks!

**Comment by kaixih** [comment]
We are thinking to put the support of LayerNorm into the library, e.g. cuDNN in the future. Then, maybe we can "hide" the TF support path in the FusedBatchNormV3, meaning the FusedBatchNormV3 will be more a generic norm (like the nn.batch_normalization which can support any shapes/broadcast). So, at this point, we might want to put this PR on hold. How do you think?

**Comment by kaixih** [comment]
We plan to resort to the cuDNN library for this functionality. And the integration will probably follow https://github.com/tensorflow/tensorflow/pull/53483#issuecomment-1049261917. Will close this PR for now.
